File: backend-clinic/src/clinic/clinic-user.service.ts
import {
  BadRequestException,
  ConflictException,
  Injectable,
  NotFoundException,
} from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { In, Repository } from "typeorm";
import { ClinicUserEntity } from "./entities/clinic-user.entity";
import { PasswordService } from "./auth/password.service";
import { randomUUID } from "crypto";

export interface ClinicUserCreateDto {
  name?: string;
  email: string;
  role?: string;
  status?: string;
  password?: string;
  mustChangePassword?: boolean;
}

export interface ClinicUserUpdateDto {
  name?: string;
  email?: string;
  avatarUrl?: string | null;
  role?: string;
  status?: string;
  mustChangePassword?: boolean;
}

export interface ClinicUserResetDto {
  password: string;
  mustChangePassword?: boolean;
}

@Injectable()
export class ClinicUserService {
  constructor(
    @InjectRepository(ClinicUserEntity)
    private readonly users: Repository<ClinicUserEntity>,
    private readonly passwordService: PasswordService,
  ) {}

  async list(tenantId: string) {
    return this.users.find({ where: { tenantId }, order: { createdAt: "DESC" } });
  }

  async listPatients(tenantId: string) {
    const rows = await this.users.find({ where: { tenantId }, order: { createdAt: "DESC" } });
    return rows.filter((user) => user.role?.toLowerCase() === "patient");
  }

  async searchPatients(tenantId: string, query: string, limit = 12) {
    const term = query.trim().toLowerCase();
    if (!term) return [];
    const like = `%${term}%`;
    return this.users
      .createQueryBuilder("u")
      .where("u.tenantId = :tenantId", { tenantId })
      .andWhere("LOWER(u.role) = 'patient'")
      .andWhere("(LOWER(u.name) LIKE :like OR LOWER(u.email) LIKE :like)", { like })
      .orderBy("u.createdAt", "DESC")
      .limit(limit)
      .getMany();
  }

  async create(tenantId: string, dto: ClinicUserCreateDto) {
    if (!dto?.email) {
      throw new BadRequestException("Email required");
    }
    const email = dto.email.trim().toLowerCase();
    const existing = await this.users.findOne({ where: { tenantId, email } });
    if (existing) {
      throw new ConflictException("Email already exists");
    }
    if (!dto.password) {
      throw new BadRequestException("Password required");
    }
    const user = this.users.create({
      id: randomUUID(),
      tenantId,
      email,
      name: dto.name ?? null,
      role: dto.role || "staff",
      status: dto.status || "active",
      passwordHash: this.passwordService.hash(dto.password),
      mustChangePassword: Boolean(dto.mustChangePassword),
      createdAt: new Date(),
      updatedAt: new Date(),
    });
    return this.users.save(user);
  }

  async update(tenantId: string, id: string, dto: ClinicUserUpdateDto) {
    const user = await this.users.findOne({ where: { tenantId, id } });
    if (!user) {
      throw new NotFoundException("User not found");
    }
    if (dto.email) {
      const email = dto.email.trim().toLowerCase();
      if (email !== user.email) {
        const exists = await this.users.findOne({ where: { tenantId, email } });
        if (exists) {
          throw new ConflictException("Email already exists");
        }
        user.email = email;
      }
    }
    if (dto.name !== undefined) user.name = dto.name ?? null;
    if (dto.avatarUrl !== undefined) user.avatarUrl = dto.avatarUrl ?? null;
    if (dto.role) user.role = dto.role;
    if (dto.status) user.status = dto.status;
    if (dto.mustChangePassword !== undefined) {
      user.mustChangePassword = Boolean(dto.mustChangePassword);
    }
    user.updatedAt = new Date();
    return this.users.save(user);
  }

  async resetPassword(tenantId: string, id: string, dto: ClinicUserResetDto) {
    if (!dto?.password) {
      throw new BadRequestException("Password required");
    }
    const user = await this.users.findOne({ where: { tenantId, id } });
    if (!user) {
      throw new NotFoundException("User not found");
    }
    user.passwordHash = this.passwordService.hash(dto.password);
    user.mustChangePassword = Boolean(dto.mustChangePassword);
    user.updatedAt = new Date();
    return this.users.save(user);
  }

  async changePassword(
    tenantId: string,
    id: string,
    currentPassword: string,
    newPassword: string,
  ) {
    if (!currentPassword || !newPassword) {
      throw new BadRequestException("Missing password");
    }
    const user = await this.users.findOne({ where: { tenantId, id } });
    if (!user) {
      throw new NotFoundException("User not found");
    }
    if (!this.passwordService.matches(currentPassword, user.passwordHash)) {
      throw new BadRequestException("Invalid password");
    }
    user.passwordHash = this.passwordService.hash(newPassword);
    user.mustChangePassword = false;
    user.updatedAt = new Date();
    return this.users.save(user);
  }

  async delete(tenantId: string, id: string) {
    const user = await this.users.findOne({ where: { tenantId, id } });
    if (!user) {
      throw new NotFoundException("User not found");
    }
    await this.users.remove(user);
  }

  async getById(tenantId: string, id: string) {
    const user = await this.users.findOne({ where: { tenantId, id } });
    if (!user) {
      throw new NotFoundException("User not found");
    }
    return user;
  }

  async listByIds(tenantId: string, ids: string[]) {
    if (!ids.length) return [];
    return this.users.find({ where: { tenantId, id: In(ids) } });
  }
}

File: backend-java/src/main/java/com/neria/manager/clinicflow/services/ClinicUserService.java
package com.neria.manager.clinicflow.services;

import com.neria.manager.auth.AuthService;
import com.neria.manager.clinicflow.entities.ClinicUser;
import com.neria.manager.clinicflow.repos.ClinicUserRepository;
import com.neria.manager.common.services.ScryptHasher;
import com.neria.manager.config.AppProperties;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class ClinicUserService {
  private final ClinicUserRepository repository;
  private final ScryptHasher hasher;
  private final AuthService authService;
  private final String salt;

  public ClinicUserService(
      ClinicUserRepository repository,
      ScryptHasher hasher,
      AuthService authService,
      AppProperties properties) {
    this.repository = repository;
    this.hasher = hasher;
    this.authService = authService;
    this.salt = Optional.ofNullable(properties.getSecurity().getClinicPasswordSalt()).orElse("");
    if (this.salt.length() < 16) {
      throw new IllegalStateException("CLINIC_PASSWORD_SALT must be at least 16 characters");
    }
  }

  public List<ClinicUserResponse> list(String tenantId) {
    return repository.findAllByTenantId(tenantId).stream()
        .sorted(Comparator.comparing(ClinicUser::getCreatedAt).reversed())
        .map(ClinicUserResponse::fromEntity)
        .toList();
  }

  public ClinicUserResponse create(String tenantId, CreateClinicUserRequest dto) {
    if (dto == null) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Missing payload");
    }
    String email = normalizeEmail(dto.email);
    if (email == null) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email required");
    }
    if (repository.findByTenantIdAndEmail(tenantId, email).isPresent()) {
      throw new ResponseStatusException(HttpStatus.CONFLICT, "Email already exists");
    }
    if (dto.password == null || dto.password.isBlank()) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Password required");
    }
    ClinicUser user = new ClinicUser();
    user.setId(UUID.randomUUID().toString());
    user.setTenantId(tenantId);
    user.setEmail(email);
    user.setName(dto.name);
    user.setRole(dto.role != null && !dto.role.isBlank() ? dto.role : "staff");
    user.setStatus(dto.status != null && !dto.status.isBlank() ? dto.status : "active");
    user.setPasswordHash(hashPassword(dto.password));
    user.setMustChangePassword(Boolean.TRUE.equals(dto.mustChangePassword));
    user.setCreatedAt(LocalDateTime.now());
    user.setUpdatedAt(LocalDateTime.now());
    return ClinicUserResponse.fromEntity(repository.save(user));
  }

  public ClinicUserResponse update(String tenantId, String id, UpdateClinicUserRequest dto) {
    ClinicUser user = repository.findByIdAndTenantId(id, tenantId)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found"));
    if (dto == null) {
      return ClinicUserResponse.fromEntity(user);
    }
    if (dto.name != null) {
      user.setName(dto.name);
    }
    if (dto.email != null && !dto.email.isBlank()) {
      String email = normalizeEmail(dto.email);
      if (!email.equalsIgnoreCase(user.getEmail())) {
        if (repository.findByTenantIdAndEmail(tenantId, email).isPresent()) {
          throw new ResponseStatusException(HttpStatus.CONFLICT, "Email already exists");
        }
        user.setEmail(email);
      }
    }
    if (dto.role != null && !dto.role.isBlank()) {
      user.setRole(dto.role);
    }
    if (dto.status != null && !dto.status.isBlank()) {
      user.setStatus(dto.status);
    }
    if (dto.mustChangePassword != null) {
      user.setMustChangePassword(dto.mustChangePassword);
    }
    user.setUpdatedAt(LocalDateTime.now());
    return ClinicUserResponse.fromEntity(repository.save(user));
  }

  public void delete(String tenantId, String id) {
    ClinicUser user = repository.findByIdAndTenantId(id, tenantId)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found"));
    repository.delete(user);
  }

  public ClinicUserResponse resetPassword(
      String tenantId, String id, ResetPasswordRequest dto) {
    if (dto == null || dto.password == null || dto.password.isBlank()) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Password required");
    }
    ClinicUser user = repository.findByIdAndTenantId(id, tenantId)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found"));
    user.setPasswordHash(hashPassword(dto.password));
    user.setMustChangePassword(Boolean.TRUE.equals(dto.mustChangePassword));
    user.setUpdatedAt(LocalDateTime.now());
    return ClinicUserResponse.fromEntity(repository.save(user));
  }

  public LoginResponse login(String tenantId, LoginRequest dto) {
    if (dto == null) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Missing payload");
    }
    String email = normalizeEmail(dto.email);
    if (email == null || dto.password == null) {
      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Email and password required");
    }
    ClinicUser user = repository.findByTenantIdAndEmail(tenantId, email)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials"));
    if (!"active".equalsIgnoreCase(user.getStatus())) {
      throw new ResponseStatusException(HttpStatus.FORBIDDEN, "User inactive");
    }
    if (!hasher.matches(dto.password, salt, user.getPasswordHash())) {
      throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Invalid credentials");
    }
    var token = authService.issueClinicToken(tenantId, user.getId(), user.getRole());
    return new LoginResponse(token.accessToken(), token.expiresIn(), ClinicUserResponse.fromEntity(user));
  }

  public ClinicUserResponse getById(String tenantId, String id) {
    ClinicUser user = repository.findByIdAndTenantId(id, tenantId)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found"));
    return ClinicUserResponse.fromEntity(user);
  }

  private String hashPassword(String value) {
    return hasher.hash(value, salt);
  }

  private String normalizeEmail(String value) {
    if (value == null) return null;
    String trimmed = value.trim().toLowerCase();
    return trimmed.isBlank() ? null : trimmed;
  }

  public static class CreateClinicUserRequest {
    public String name;
    public String email;
    public String role;
    public String status;
    public String password;
    public Boolean mustChangePassword;
  }

  public static class UpdateClinicUserRequest {
    public String name;
    public String email;
    public String role;
    public String status;
    public Boolean mustChangePassword;
  }

  public static class ResetPasswordRequest {
    public String password;
    public Boolean mustChangePassword;
  }

  public static class LoginRequest {
    public String email;
    public String password;
    public String tenantId;
    public String serviceCode;
  }

  public record ClinicUserResponse(
      String id,
      String tenantId,
      String name,
      String email,
      String role,
      String status,
      boolean mustChangePassword,
      LocalDateTime createdAt,
      LocalDateTime updatedAt) {
    public static ClinicUserResponse fromEntity(ClinicUser entity) {
      return new ClinicUserResponse(
          entity.getId(),
          entity.getTenantId(),
          entity.getName(),
          entity.getEmail(),
          entity.getRole(),
          entity.getStatus(),
          entity.isMustChangePassword(),
          entity.getCreatedAt(),
          entity.getUpdatedAt());
    }
  }

  public record LoginResponse(String accessToken, long expiresIn, ClinicUserResponse user) {}
}
